Lecture notes for week 1 (#1)
HarrisonGrodin/ATPL24: week1.tex

denoted $\hat{\gamma}$.

\begin{align*}
  \hat{\gamma}(x) &= \gamma(x) \\
  \hat{\gamma}(\unitEx*) &= \unitEx* \\
  \hat{\gamma}(\pairEx*{M_1}{M_2}) &= \pairEx*{\hat{\gamma}(M_1)}{\hat{\gamma}(M_2)} \\
  \hat{\gamma}(\fstEx*{M}) &= \fstEx*{\hat{\gamma}(M)} \\
  \hat{\gamma}(\sndEx*{M}) &= \sndEx*{\hat{\gamma}(M)} \\
  \hat{\gamma}(\lamEx*{x}{M}) &= \lamEx*{x}{\hat{\gamma}
  '(M)} \tag*{where $\gamma' = \gamma\{x \mapsto x\}$} \\
  \hat{\gamma}(\appEx*{M_1}{M_2}) &= \appEx*{\hat{\gamma}(M_1)}{\hat{\gamma}(M_2)} \\
  \hat{\gamma}(\yesEx*) &= \yesEx* \\
  \hat{\gamma}(\noEx*) &= \noEx*
\end{align*}

The substitution lemma that governs it is of essence.

\begin{lemma}\label{lem:subst}
If $\Gamma \entails \isOfTp{M}{A}$ and $\gamma : \Gamma' \to \Gamma$ then
$\Gamma' \entails \isOfTp{\hat{\gamma}(M)}{A}$ \footnote{In a dependently-typed context, the substitution also applies to $A$, then the conclusion
is $\Gamma' \entails \isOfTp{\hat{\gamma}(M)}{\hat{\gamma}(A)}$.}.
\end{lemma}

The term $\hat{\gamma}(M)$ is commonly abbreviated as $\hat{M}$ when the substitution $\gamma$ applied
is clear from context.

> \begin{corollary}[Transitivity/Cut/Substituion]

Reviewer: ```suggestion
\begin{corollary}[Transitivity/Cut/Substitution]
```

not necessarily a value, and we do not have information on its evaluation either.
We are again stuck\footnote{The reader may attempt to salvage the proof at this point by a term size argument,
or switching the evaluation to by-value. The next case would reveal why it is hopeless.}.

The situation is more dire for the case of \ruleref{sta:app}. In this case the IH tells us
the principal argument $M_1 \evalsTo \lamEx*{x}{M_1'}$ (and $M_2 \evalsTo v$ for some $v$).
Therefore $\appEx*{M_1}{M_2} \stepsTo(*) \appEx*{\lamEx*{x}{M_1'}}{M_2} \stepsTo \Sub{M_2}{x}{M_1'}$
\footnote{The term resulting from substitution could be (and usually is) much larger in size compared to the original term.}.
We are stuck.

\paragraph{Attempt 3.} It is clear from our previous proofs that we need stronger invariants on the
values resulting from execution. This motivates us to \emph{define} a family of \emph{predicates}
\footnote{In other words, $\HT[\cdot] \colon \Opn{Tp} \to (\Opn{Tm} \to \Opn{prop})$, and
the motive for the induction is the type $\Opn{Tm} \to \Opn{prop}$.} $\HT[A](M)$ over
\emph{closed} terms $M$, indexed by types $A$.

> \begin{definition}[Hereidtary Termination]

Reviewer: ```suggestion
\begin{definition}[Hereditary Termination]
```

\begin{mathpar}
\defrule[V-$\natTy*$-1][dyn:val-zero]
  {\strut}
  {\isVal{\zeroEx*}}

\defrule[V-$\natTy*$-2][dyn:val-succ]
  {\strut}
  {\isVal{\succEx*{M}}}
\end{mathpar}

\begin{mathpar}

\defrule[D-app-c][dyn:nat-c]
  {M \stepsTo{} M'}
  {\natitEx*{M}{M_z}{x}{M_s} \stepsTo{} \natitEx*{M'}{M_z}{x}{M_s}}

\defrule[D-natit-1][dyn:natit-zero]
  {\strut}
  {\natitEx*{\zeroEx*}{M_z}{x}{M_s} \stepsTo{} M_z}

\defrule[D-natit-2][dyn:natit-succ]
  {\strut}
  {\natitEx*{\succEx*{M}}{M_z}{x}{M_s} \stepsTo{} \Sub{\natitEx*{M}{M_z}{x}{M_s}}{x}{M_s}}
\end{mathpar}

Again, our goal is to prove termination for the calculus with natural numbers. Following the development in \Cref{sec:standard}, we proceed by first defining $\HT[\natTy*](-)$.

\begin{definition}[Hereidtary Termination for $\natTy*$]\label{def:ht-nat}
  Define $\HT[\natTy*](M)$ to the \emph{strongest} predicate
  \footnote{When viewed as sets, the smallest set of terms.} $\mathcal{P}$ on terms $M$ such that one of the following is true:
\begin{enumerate}
  \item If $M \evalsTo \zeroEx*$, then $\mathcal{P}(M)$.
  \item If $M \evalsTo \succEx*{M'}$ and $\mathcal{P}(M')$, then $\mathcal{P}(M)$.
\end{enumerate}
\end{definition}

> Let us pondering on this definition and think about what it entails.

Reviewer: ```suggestion
Let us ponder this definition and think about what it entails.
```
